runhouse/resources/functions/aws_lambda.py

```diff
@@ -0,0 +1,887 @@
+from typing import Any, Callable, List, Optional
+
+import boto3
+import botocore.exceptions
+
+from runhouse.globals import rns_client
+from runhouse.resources.envs import _get_env_from, Env
+
+from runhouse.resources.functions.function import Function
+
+logger = logging.getLogger(__name__)
+
+CRED_PATH = f"{Path.home()}/.aws/credentials"
+SUPPORTED_RUNTIMES = [
+    "python3.7",
+    "python3.8",
+    "python3.9",
+    "python3.10",
+    "python 3.11",
+]
+DEFAULT_PY_VERSION = "python3.9"
+
+
+class LambdaFunction(Function):
+    RESOURCE_TYPE = "lambda_function"
+    DEFAULT_ACCESS = "write"
+    DEFAULT_ROLE_POLICIES = [
+        "cloudwatch:*",
+        "lambda:Invoke",
+        "lambda:InvokeAsync",
+        "lambda:InvokeFunction",
+        "lambda:PublishVersion",
+        "logs:*",
+        "s3:DeleteObject",
+        "s3:GetObject",
+        "s3:ListBucket",
+        "s3:PutObject",
+    ]
+    MAX_WAIT_TIME = 60  # seconds, max time that can pass before we raise an exception that AWS update takes too long.
+    DEFAULT_REGION = "us-east-1"
+    DEFAULT_RETENTION = 30  # one month, for lambdas log groups.
+    DEFAULT_TIMEOUT = 900  # sec. meaning 15 min.
+    DEFAULT_MEMORY_SIZE = 1024  # MB
+    DEFAULT_TMP_SIZE = 3072  # MB, meaning 3G
+    HOME_DIR = "/tmp/home"
+
+    def __init__(
+        self,
+        paths_to_code: list[str],
+        handler_function_name: str,
+        runtime: str,
+        args_names: list[str],
+        fn_pointers: tuple,
+        name: str,
+        env: Env,
+        timeout: int,  # seconds
+        memory_size: int,  # MB
+        tmp_size: int,  # MB
+        retention_time: int = DEFAULT_RETENTION,  # days
+        dryrun: bool = False,
+        function_arn: Optional[str] = None,
+        access: Optional[str] = None,
+        **kwargs,  # We have this here to ignore extra arguments when calling from from_config
+    ):
+        """
+        Runhouse AWS Lambda object. It is comprised of the entry point, configuration,
+        and dependencies necessary to run the service on AWS infra.
+
+        .. note::
+                To create an AWS lambda resource, please use the factory method :func:`lambda_function`.
+        """
+
+        super().__init__(
+            name=name,
+            dryrun=dryrun,
+            system=self.RESOURCE_TYPE,
+            env=env,
+            fn_pointers=fn_pointers,
+            access=access or self.DEFAULT_ACCESS,
+            **kwargs,
+        )
+
+        self.local_path_to_code = paths_to_code
+        self.handler_function_name = handler_function_name
+        self.args_names = args_names
+        self.runtime = runtime
+        self.timeout = timeout
+        self.memory_size = memory_size
+        self.tmp_size = tmp_size
+        self.aws_lambda_config = (
+            None  # Lambda config and role arn from aws will be saved here
+        )
+        self.retention_time = retention_time
+        self.function_arn = function_arn
```

Comment:
In what scenario would `function_arn` be passed in as an argument?

Reply:
It is saved in DEN, and used for sharing a lambda: if user A in region X shares a lambda with user B in region Y, using the saved arn, user B will be able to load and invoke the function.